lib/metadata/mongoclient/backend.js

```diff
@@ -0,0 +1,554 @@
+ * may generate orphans but not alter the proper conduct of operations
+ * (what he user wants and what we acknowledge to the user).
+ *
+ * Orphan situations may be recovered by the Lifecycle.
+ *
+ * We use proper atomic operations when needed.
+ */
+
+const util = require('util');
+const arsenal = require('arsenal');
+
+const logger = require('../../utilities/logger');
+
+const constants = require('../../../constants');
+const { config } = require('../../Config');
+
+const errors = arsenal.errors;
+const versioning = arsenal.versioning;
+const BucketInfo = arsenal.models.BucketInfo;
+
+const MongoClient = require('mongodb').MongoClient;
+
+const Version = versioning.Version;
+const genVID = versioning.VersionID.generateVersionId;
+
+const MongoReadStream = require('./readStream');
+
+const METASTORE = '__metastore';
+
+let uidCounter = 0;
+
+const VID_SEP = versioning.VersioningConstants.VersionId.Separator;
+
+function generateVersionId() {
+    // generate a unique number for each member of the nodejs cluster
+    return genVID(`${process.pid}.${uidCounter++}`,
+		  config.replicationGroupId);
+}
+
+function formatVersionKey(key, versionId) {
+    return `${key}${VID_SEP}${versionId}`;
+}
+
+function inc(str) {
+    return str ? (str.slice(0, str.length - 1) +
+            String.fromCharCode(str.charCodeAt(str.length - 1) + 1)) : str;
+}
+
+const VID_SEPPLUS = inc(VID_SEP);
+
+function generatePHDVersion(versionId) {
+    return { "isPHD": true, "versionId": versionId };
+}
+
+class MongoClientInterface {
+    constructor() {
+	const mongoUrl =
+	      `mongodb://${config.mongodb.host}:${config.mongodb.port}`;
+
+	this.client = null;
+	this.db = null;
+	console.log('connecting to', mongoUrl);
+        MongoClient.connect(mongoUrl, (err, client) => {
+	    if (err) {
+		throw (errors.InternalError);
+	    }
+	    console.log('***CONNECTED TO MONGODB***');
+	    this.client = client;
+	    this.db = client.db(config.mongodb.database);
+	    this.usersBucketHack();
+	});
+    }
+
+    usersBucketHack() {
+	/* Since the bucket creation API is expecting the usersBucket
+           to have attributes, we pre-create the usersBucket
+           attributes here (see bucketCreation.js line 36)*/
+        const usersBucketAttr = new BucketInfo(constants.usersBucket,
+            'admin', 'admin', new Date().toJSON(),
+            BucketInfo.currentModelVersion());
+	this.createBucket(
+            constants.usersBucket,
+            usersBucketAttr, {}, err => {
+                if (err) {
+                    console.log('error writing usersBucket ' +
+                                'attributes to metastore',
+                                { error: err });
+                    throw (errors.InternalError);
+                }
+            });
+    }
+    
+    getCollection(name) {
+	/* mongo has a problem with .. in collection names */
+	if (name === constants.usersBucket)
+	    name = "users__bucket";
+	return this.db.collection(name);
```

Comment:
do we want to use the strict version with a callback so that we can provide an error if the collection does not exist or handle appropriately? Skip comment. As discussed, we are ok having the collection created on a put object.